### DISCUSSION TITLE
Layers got deleted Automatically on saving PDF page.

### DISCUSSION THREAD

**Discussion by pachhaipurna** [pymupdf/PyMuPDF#2767] [Q&A] [answered]
@JorjMcKie` I am creating new pdf page based on page number , While creating new pdf page all annotation attribute are correct except Layer.Layers got automatically removed while saving pdf page. This is my code.

`output_pdf_with_color = fitz.open(filetype = "pdf")
input_pdf_page = fitz.open(input_src_path,filetype = "pdf")
for page_no in pdf_markup_rslt[subsystem_key]['Page No']:
    page_no_in_str = page_no_in_str + ' Pages ' + str(page_no)
    output_pdf_with_color.insert_pdf(input_pdf_page,from_page = page_no-1, to_page = page_no-1, links=1, annots=1)

    pdf_with_color = os.path.join(subsystem_dir, '{}_final.pdf'.format(subsystem_key))
    output_pdf_with_color.save(pdf_with_color)`

Could you please let me know How to fix it ?

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/b716c0da-f00e-441b-b7f8-a8b66d208aab)

Original file contains layers.

Thank you

**Comment by JorjMcKie** [comment]
For the time being, this is not a bug, but a question.
Transferring to "Discussions".

**Comment by JorjMcKie** [comment, marked as answer]
I am sorry, but your code is hardly readable, because of the mixture formatted and unformatted code.
What I understood is that your output PDF is filled with pages from some source PDF via `insert_pdf()`.
Correct so far?

**Reply by pachhaipurna** [reply]
@JorjMcKie Yes, I want to add specific page number to new pdf page using **insert_pdf()**, While saving to pdf output all Layers that are present originally are removed, I am unable to save layers in my output file. Your help is highly appreciated.

  ` ###Output file
    output_pdf_with_color = fitz.open(filetype = "pdf")

    input_pdf_page = fitz.open(input_src_path,filetype = "pdf")
    ##Page Number to be prepared,response data
    
    for page_no in pdf_markup_rslt[subsystem_key]['Page No']:
        page_no_in_str = page_no_in_str + ' Pages ' + str(page_no)

        output_pdf_with_color.insert_pdf(input_pdf_page,from_page = page_no-1, to_page = page_no-1, links=1, annots=1)

        pdf_with_color = os.path.join(subsystem_dir, '{}_final.pdf'.format(subsystem_key))
        output_pdf_with_color.save(pdf_with_color)`

Thank you.

**Comment by JorjMcKie** [comment]
Inserting pages from a source to a target PDF is a **_page-oriented_** process. Layers however are stored centrally in a PDF - not by page or inside a page.

While OCGs referenced by a page will copied over to the targt PDF - like any other referenced object - they will still not become known as OCGs by this copy.

You therefore must find a way to replicate the source **_Optional Content situation_** in the target somehow.

This is currently not supported and requires your own - probably hacky - code. I cannot think of a complete and failsafe way to do that to be honest.

**Reply by pachhaipurna** [reply]
Thank you @JorjMcKie for your response.

**Reply by pachhaipurna** [reply]
@JorjMcKie I have updated the code as below,
My pdf has lots of pages more than 200 -300 pages in each pdf.

  `import fitz
from fitz import utils

###Processing function for PDF.
def process_pdf_page():

    input_src_path = 'C:/input_pdf_src.pdf' #### Large pdf file more than 200 MB,
    input_pdf_page = fitz.open(input_src_path,filetype = "pdf")
    last_page = input_pdf_page.page_count - 1
    print('Original Length of Xref :: ',input_pdf_page.xref_length() )

    mov_idx = 0
    pagelist = [1,3,5,7,19], ###This will be dynamic 
    for page_no in pagelist:

        if page_no-1 == mov_idx:
            print('No Need to change location/Move')
        else:
            input_pdf_page.move_page(page_no-1, mov_idx)

    input_pdf_page.delete_pages(from_page = len(pagelist), to_page=last_page)
    print('Updated Xref Length :: ',input_pdf_page.xref_length() )

    input_pdf_page.save('Output.pdf', clean = True, deflate=False)
    input_pdf_page.close()

if __name__ == '__main__':
    process_pdf_page()`

The issue is file size is not decreased even if I delete more than half of the pages, I used garbage =3 while saving but no any effect in the new pdf file that is created, Could you please help me on this?
Thank you.

**Reply by JorjMcKie** [reply]
You must use "garbage=3, deflate=True, clean=False".
Even then it is possible that the file size does not decrease (significantly).
It is wrong to assume that the PDF size is in linear correlation with the number of pages.
Deleting pages in the first place just means that the reference to the page object is removed from the page tree. Nothing more.
It does not guarantee that the page will also be physically removed even if garbage collection is requested: there may exist references in other places ... For example, the page may be referred to by so-called "article beads".
Things like this will cause the page to survive as a "ghost" object and no space reduction will be seen.

**Reply by pachhaipurna** [reply]
@JorjMcKie  I tried garbage=3, deflate=True, clean=False". but it takes lots of time, Could you please let me know other process from pymupdf?

**Reply by JorjMcKie** [reply]
Try reducing the garbage value to 2 or 1.

**Reply by pachhaipurna** [reply]
@JorjMcKie Already tried but no any effect .Is there any option we can delete ghost object. or while deleting page we can delete those objects permanently.

**Reply by pachhaipurna** [reply]
@JorjMcKie PDF file size is increasing than the original one, Even though I have deleted more than half of the page.

**Reply by JorjMcKie** [reply]
You could try one last thing: To remove possible "article beads" in the source PDFdo this:
```python
import fitz
src = fitz.open(source.pdf")
for page in src:
    pxref = page.xref
    src.xref_set_key(pxref, "B", "null")  # remove any reference to article beads

# no execute the page insertions to the target
doc = fitz.open(...)  # your target PDF
doc.insert_pdf(src, ...)
doc.ez_save(...)
```
